crates/storage/src/schema/revision_0071.rs

```diff
@@ -0,0 +1,477 @@
+//! Prior to this migration, [state update](crate::connection::state_update)
+//! related tables have had FOREIGN KEY references with the ON DELETE CASCADE
+//! action to the `canonical_blocks` table. The goal was to make purging blocks
+//! easier by having a cascading delete effect on these tables when a row from
+//! `canonical_blocks` is deleted.
+//!
+//! With the introduction of [blockchain pruning](crate::connection::pruning), a
+//! need has arisen to delete block related table entries (including
+//! `canonical_blocks`) independently of the state update tables. This was, of
+//! course, not possible due to the aforementioned FOREIGN KEY references
+//! leading to multiple workarounds and unnecessary complexity.
+//!
+//! This migration removes the FOREIGN KEY references to `canonical_blocks` from
+//! the state update tables.
+
+use anyhow::Context;
+use pathfinder_common::BlockNumber;
+use rusqlite::params;
+
+const LOG_INTERVAL: std::time::Duration = std::time::Duration::from_secs(5);
+
+struct TableUpdate {
+    table_name: &'static str,
+    create_table_stmt: &'static str,
+    transfer_fn: transfer::TransferFn,
+    drop_and_rename_stmt: &'static str,
+    create_indices_fn: indices::CreateIndicesFn,
+}
+
+pub(crate) fn migrate(tx: &rusqlite::Transaction<'_>) -> anyhow::Result<()> {
+    tracing::info!("Dropping FOREIGN KEY references from several state update related tables");
+
+    let table_updates = [
+        TableUpdate {
+            table_name: "block_signatures",
+            create_table_stmt: r"
+                CREATE TABLE block_signatures_new(
+                    block_number INTEGER REFERENCES block_headers(number) ON DELETE CASCADE,
+                    signature_r  BLOB NOT NULL,
+                    signature_s  BLOB NOT NULL
+                );
+            ",
+            transfer_fn: transfer::block_signatures,
+            drop_and_rename_stmt: r"
+                DROP TABLE block_signatures;
+                ALTER TABLE block_signatures_new RENAME TO block_signatures;
+            ",
+            create_indices_fn: indices::block_signatures,
+        },
+        TableUpdate {
+            table_name: "class_definitions",
+            create_table_stmt: r"
+                CREATE TABLE class_definitions_new(
+                    hash         BLOB PRIMARY KEY,
+                    definition   BLOB,
+                    block_number INTEGER REFERENCES block_headers(number) ON DELETE SET NULL
```

Comment:
We should probably remove this foreign key reference too.

The problem is that when the block header gets pruned we lose the information where the class has been declared, which might lead to wrong results when executing on old state.

Reply:
Yes, that's what I've meant.